lib/elrequest.js

```diff
@@ -0,0 +1,187 @@
+'use strict'
+
+// A tiny "El"astic HTTP "request" library.
+
+const { URL } = require('url')
+
+// A wrapper around `{http|https}.request()` that adds support for a connection
+// timeout separate from the existing `options.timeout`.
+//
+// The existing `options.timeout` to `http.request()` sets `socket.setTimeout()`
+// which will emit the 'timeout' event if there is an period of socket idleness
+// that is this long. In practice for short-lived requests, it is a timeout on
+// getting the start of response data back from the server.
+//
+// The new `opts.connectTimeout` is a number of milliseconds count from socket
+// creation to socket 'connect'. If this time is reached a 'connectTimeout'
+// event will be emitted on the request object. As with 'timeout', it is up
+// to the caller to handle destroying the request. See "Usage" below.
+// In pratice this allows for a shorter timeout to see if the remote server
+// is handling connections in a timely manner. To be useful, a `connectTimeout`
+// is typically shorter than a given `timeout`.
+//
+// Note on call signature:
+//
+// Currenly this only supports the one signature:
+//      elRequest(url: String, opts?: Object, cb?: Function)
+// rather than the three signatures (and 25 lines of argument normalization)
+// that core `http.request()` supports. If we need the other signatures we
+// can add them later.
+//
+// Usage:
+//     const { elRequest } = require('./elrequest')
+//
+//     var req = elRequest(url, {
+//         connectTimeout: connectTimeout,
+//         // Any {http|https}.request options...
+//         timeout: timeout
+//     }, function onRes(res) {
+//         // Handling of the response...
+//     })
+//
+//     req.on('timeout', function () {
+//         // ...
+//         req.destroy()
+//     });
+//
+//     req.on('connectTimeout', function () {
+//         // ...
+//         req.destroy()
+//     });
+//
+//     req.on('error', function (err) {
+//         // ...
+//     })
+//
+//     req.end()
+//
+function elRequest (url, opts, cb) {
+  const {
+    connectTimeout,
+    ...reqOpts
+  } = opts || {}
+
+  // http or https
+  const protocol = reqOpts.protocol || (new URL(url)).protocol
+  let proto
+  if (protocol === 'http:') {
+    proto = require('http')
+  } else if (protocol === 'https:') {
+    proto = require('https')
+  } else {
+    throw new Error(`unsupported protocol: "${protocol}"`)
+  }
+
+  const req = proto.request(url, reqOpts, cb)
+
+  if (connectTimeout) {
+    // Handle a connection timeout with a timer starting when the request
+    // socket is *created* ("socket" event) and ending when the socket
+    // is connected.
+    req.on('socket', function (socket) {
+      // log.trace({url: url}, 'start connectTimeout')
+      var connectTimer = setTimeout(function onConnectTimeout () {
+        // log.trace({url: url}, 'connectTimeout')
+        req.emit('connectTimeout')
```

Comment:
Should we also send along an error object with the `connectTimeout`?
```
const error = new Error('some useful message about the socket timing out')
req.emit('connectTimeout', error) 
```

or is that better left to the client listening for the `connectTimeout` event?

Reply:
Node core does not include any event data with the "timeout event: https://github.com/nodejs/node/blob/948f5152d7e90451439cb8040fd15237decbc229/lib/_http_client.js#L723
I am inclined to do the same.